core/src/main/java/ai/timefold/solver/core/impl/solver/termination/UnimprovedBestSolutionTermination.java

```diff
@@ -0,0 +1,178 @@
+    // this criterion is specifically used to identify flat lines among multiple curves before the termination.
+    // The goal is to adjust the stop criterion based on the latest curve found when there are several.
+    private final double minUnimprovedBestSolutionLimit;
+    // The field stores the first best solution move count found in the curve growth chart.
+    // If a solving process involves multiple curves,
+    // the value is tied to the growth of the last curve analyzed.
+    protected long initialImprovementMoveCount;
+    protected long lastImprovementMoveCount;
+    protected long lastMoveEvaluationSpeed;
+    private Score<?> previousBest;
+    protected Score<?> currentBest;
+    protected boolean waitForFirstBestScore;
+    protected Boolean terminate;
+
+    public UnimprovedBestSolutionTermination(double unimprovedBestSolutionLimit) {
+        this.maxUnimprovedBestSolutionLimit = unimprovedBestSolutionLimit;
+        // 80% of the max unimproved limit
+        this.minUnimprovedBestSolutionLimit = unimprovedBestSolutionLimit * 0.8;
+        if (unimprovedBestSolutionLimit < 0) {
+            throw new IllegalArgumentException(
+                    "The unimprovedBestSolutionLimit (%.2f) cannot be negative.".formatted(unimprovedBestSolutionLimit));
+        }
+    }
+
+    public double getUnimprovedBestSolutionLimit() {
+        return maxUnimprovedBestSolutionLimit;
+    }
+
+    // ************************************************************************
+    // Lifecycle methods
+    // ************************************************************************
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void phaseStarted(AbstractPhaseScope<Solution_> phaseScope) {
+        super.phaseStarted(phaseScope);
+        initialImprovementMoveCount = 0L;
+        lastImprovementMoveCount = 0L;
+        lastMoveEvaluationSpeed = 0L;
+        currentBest = phaseScope.getBestScore();
+        previousBest = currentBest;
+        waitForFirstBestScore = true;
+        terminate = null;
+    }
+
+    @Override
+    public void stepStarted(AbstractStepScope<Solution_> stepScope) {
+        super.stepStarted(stepScope);
+        terminate = null;
+    }
+
+    @Override
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public void stepEnded(AbstractStepScope<Solution_> stepScope) {
+        super.stepEnded(stepScope);
+        if (waitForFirstBestScore) {
+            waitForFirstBestScore = ((Score) currentBest).compareTo(stepScope.getScore()) >= 0;
+        }
+    }
+
+    // ************************************************************************
+    // Terminated methods
+    // ************************************************************************
+
+    @Override
+    public boolean isSolverTerminated(SolverScope<Solution_> solverScope) {
+        throw new UnsupportedOperationException(
+                "%s can only be used for phase termination.".formatted(getClass().getSimpleName()));
```

Comment:
I understand why this is, but as a future improvement, we should think about ways of making this termination the default without the user having to configure anything.
I want this to eventually be the only termination the user will ever need.

This will require the termination to _somehow_ survive the CH phase.